### PR TITLE
#230 - Refactor solve

### LINKED ISSUE #230: Refactor `solve.jl`
See https://github.com/JuliaReach/Reachability.jl/pull/228#discussion_r211182885

### PR DESCRIPTION
**Pull request by mforets**
Closes #230.

### CODE REVIEW

File: src/ReachSets/reach.jl [hunk 1]
```diff
@@ -214,3 +214,90 @@ function reach(S::AbstractSystem,
...
+    end
+    
+
+    reach(Δ, N, algorithm=algorithm, ε_init=ε_init, set_type_init=set_type_init,
+          ε_iter=ε_iter, assume_sparse=assume_sparse, assume_homogeneous=assume_homogeneous,
+          numeric_type=numeric_type, lazy_X0=lazy_X0, kwargs...)
+end
+
+
+function matrix_conversion(Δ, options)
```
Comment (schillic) [inline src/ReachSets/reach.jl]: :+1: for wrapping in a function

File: src/ReachSets/reach.jl [hunk 2]
```diff
@@ -214,3 +214,90 @@ function reach(S::AbstractSystem,
...
+        # set new matrix
+        if method_exists(inputset, Tuple{typeof(Δ.s)})
+            Δ = DiscreteSystem(A, Δ.x0, inputset(Δ))
+        else
+            Δ = DiscreteSystem(A, Δ.x0)
+        end
+    end
+    return Δ
+end
+
```
Comment (schillic) [inline src/ReachSets/reach.jl]: new line

File: src/solve.jl [hunk 3]
```diff
@@ -77,197 +102,110 @@ A solution object whose content depends on the input options.
...
-        (Δ, transformation_matrix) = transform(Δ, options[:coordinate_transformation])
-        tocc()
-    else
-        transformation_matrix = nothing
+#=
+    if :algorithm ∉ keys(options)
+        options[:algorithm] = default_algorithm(system)
     end
+=#
+    println(algorithm)
```
Comment (schillic) [inline src/solve.jl]: You have to decide between `algorithm` and `options[:algorithm]`. Only use one of them.

File: src/solve.jl [hunk 4]
```diff
@@ -313,3 +251,33 @@ project(reach_sol::AbstractSolution) = project(reach_sol.Xk, reach_sol.options)
 
 project(Rsets::Vector{<:LazySet}, options::Pair{Symbol,<:Any}...) =
     project(Rsets, Options(Dict{Symbol,Any}(options)))
+
+# ===========================================================================
+# Bogomolov, Foretes, Frehse, Podolski, Schillic, Viry 2018
```
Comment (schillic) [inline src/solve.jl]: Bogomolov, Forets, Frehse, Podelski, Schilling, Viry 2018

File: src/solve.jl [hunk 5]
```diff
@@ -313,3 +251,33 @@ project(reach_sol::AbstractSolution) = project(reach_sol.Xk, reach_sol.options)
...
+# ===========================================================================
+function init_BFFPSV18!(system, options)
+     # state dimension for (purely continuous or purely discrete systems)
```
Comment (schillic) [inline src/solve.jl]: The indentation here has 5 spaces.

**Comment by schillic** [comment]
The build error is `LoadError: UndefVarError: solve not defined`.

File: src/solve.jl (hunk 4 above)
Comment (mforets) [reply src/solve.jl]: that'a a typo in my own name 😄

**Comment by schillic** [comment]
*bump*

File: src/ReachSets/reach.jl [hunk 6]
```diff
@@ -214,3 +214,90 @@ function reach(S::AbstractSystem,
...
+               assume_sparse=true,
+               assume_homogeneous=false,
+               numeric_type::Type=Float64,
+               lazy_X0=false,
+               kwargs...)::Vector{<:LazySet}
+
+    # ===================
+    # Time discretization
+    # ===================
+    if system isa InitialValueProblem{<:AbstractContinuousSystem}
```
[the remark this answers is not in the document]
Comment (mforets) [reply src/ReachSets/reach.jl]: probably  here comes `S`

**Comment by schillic** [comment]
Finally, the unit tests pass. Now I have to check the benchmarks.

**Comment by schillic** [comment]
Ready from my side!

**Pull request merged**